ompi/mpi/c/reduce_local.c

```diff
@@ -76,8 +79,12 @@ int MPI_Reduce_local(const void *inbuf, void *inoutbuf, int count,
     /* Invoke the op component to perform the back-end operation */
     OBJ_RETAIN(op);
     /* XXX -- CONST -- do not cast away const -- update mca/coll */
-    ompi_op_reduce(op,(void *) inbuf, inoutbuf, count, datatype);
+    // Since there is no 'comm' parameter to this interface use 'self' since
+    // this is a local operation to this process.
+    ompi_communicator_t *comm = &ompi_mpi_comm_self.comm;
+    err = comm->c_coll.coll_reduce_local(inbuf, inoutbuf, count, datatype, op,
+                                         comm->c_coll.coll_reduce_local_module);
     OBJ_RELEASE(op);
 
-    return MPI_SUCCESS;
+    OMPI_ERRHANDLER_RETURN(err, MPI_COMM_WORLD, err, FUNC_NAME);
```

Comment:
MPI 3.1 section 8.3 page 340 line 21, specify that MPI calls not attached to an MPI object (communicator, file or window) are attached to MPI_COMM_WORLD. Basically, all error handling for such call are required to be raised in the context of MPI_COMM_WORLD.

@gpaulsen is right about the retain/release. However, the right approach will be to do it in the API and not in the actual implementation (such that when we use the function internally we don't waste time in unnecessary retain/release).